src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/ProjectSystem/DocumentContextExtensions.cs

```diff
@@ -27,13 +33,25 @@ public static async Task<Document> GetGeneratedDocumentAsync(this VersionedDocum
         var projectKey = razorDocument.Project.ToProjectKey();
         var generatedFilePath = filePathService.GetRazorCSharpFilePath(projectKey, razorDocument.FilePath.AssumeNotNull());
         var generatedDocumentId = solution.GetDocumentIdsWithFilePath(generatedFilePath).First(d => d.ProjectId == razorDocument.Project.Id);
-        var generatedDocument = solution.GetDocument(generatedDocumentId).AssumeNotNull();
+        generatedDocument = solution.GetDocument(generatedDocumentId).AssumeNotNull();
 
         var csharpSourceText = await documentContext.GetCSharpSourceTextAsync(cancellationToken).ConfigureAwait(false);
 
         // HACK: We're not in the same solution fork as the LSP server that provides content for this document
         generatedDocument = generatedDocument.WithText(csharpSourceText);
 
-        return generatedDocument;
+        // Obviously this lock is not sufficient to avoid wasted work, but it does at least avoid mutating the snapshot
+        // any more than just a once of caching of the generated document, which is what is really happening with the set
+        // method call below.
+        lock (snapshot)
+        {
+            if (snapshot.TryGetGeneratedDocument(out var generatedDocument2))
+            {
+                return generatedDocument2;
+            }
+
+            snapshot.SetGeneratedDocument(generatedDocument);
+            return generatedDocument;
+        }
```

Comment:
Consider moving this to `RemoteDocumentSnapshot` as `GetOrAddGeneratedDocument(...)`

Reply:
I've considered it, but I like this way better. Feels somehow safer to have the SetGeneratedDocument method that can throw if it's abused. But also I really hope to just delete all of this soon anyway :)

Reply:
No worries. So long as SetGeneratedDocument(...)` couldn't be called by a different thread without first locking on the `snapshot`, there won't be a problem. Normally, I prefer for objects to keep their own state consistent rather than having to lock on their instance when touching them since it's less error prone. But, if there's no chance of that and the code is temporary, I trust you.

Reply:
Obviously there isn't no chance, but I am 99% confident that if we ever see another call to `SetGeneratedDocument` in any PR, it's a mistake.